File: go/processor/extractor/external_dep.go
// @docspec:module {
//   id: "docspec-go-extractor-external-dep",
//   name: "External Dependency Extractor",
//   description: "Detects external service and API calls via net/http, resty, go-retryablehttp, and gRPC client usage, extracting URLs, methods, and call sites.",
//   since: "3.0.0"
// }
package extractor

import (
	"go/ast"
	"strings"
)

// ExternalDepExtractor detects external service/API calls in Go source code:
// @docspec:boundary "classpath-safe extraction"
//   - net/http client usage (http.Get, http.Post, http.NewRequest)
//   - Popular HTTP client libraries (resty, req, go-retryablehttp)
//   - gRPC client connections (grpc.Dial, grpc.NewClient)
//   - Comment-based docspec tags (//docspec:external)
//
// This is the Go equivalent of the Java ExternalDependencyExtractor which detects
// RestTemplate, WebClient, RestClient, and @FeignClient usage.
type ExternalDepExtractor struct{}

// Known Go HTTP/gRPC client import paths.
var externalDepImports = []string{
	"net/http",
	"github.com/go-resty/resty/v2",
	"github.com/imroc/req/v3",
	"github.com/hashicorp/go-retryablehttp",
	"google.golang.org/grpc",
}

// @docspec:deterministic
func (e *ExternalDepExtractor) Name() string {
	return "external-dependency"
}

// IsAvailable reports true if any file imports a known HTTP or gRPC client package.
//
// @docspec:deterministic
func (e *ExternalDepExtractor) IsAvailable(ctx *ProcessorContext) bool {
	for _, imp := range externalDepImports {
		if hasImport(ctx, imp) {
			return true
		}
	}
	return false
}

// Extract scans files for external dependency patterns and populates ctx.ExternalDeps.
//
// @docspec:method { since: "3.0.0" }
// @docspec:intentional "Detects HTTP client calls, resty chain calls, gRPC connections, and client struct fields for external dependency mapping"
func (e *ExternalDepExtractor) Extract(ctx *ProcessorContext) {
	seen := make(map[string]bool)

	for _, file := range ctx.Files {
		imports := fileImports(file)
		pkg := file.Name.Name

		hasHTTP := imports["net/http"]
		hasResty := imports["github.com/go-resty/resty/v2"]
		hasGRPC := imports["google.golang.org/grpc"]

		for _, decl := range file.Decls {
			fn, ok := decl.(*ast.FuncDecl)
			if !ok {
				continue
			}

			owner := pkg + "." + fn.Name.Name
			if fn.Recv != nil && len(fn.Recv.List) > 0 {
				owner = pkg + "." + recvTypeName(fn.Recv.List[0].Type) + "." + fn.Name.Name
			}

			// Check for docspec:external comment tags
			if fn.Doc != nil {
				for _, c := range fn.Doc.List {
					text := strings.TrimSpace(strings.TrimPrefix(c.Text, "//"))
					if strings.HasPrefix(text, "docspec:external") {
						parts := strings.Fields(text)
						if len(parts) >= 2 {
							name := parts[1]
							baseURL := ""
							if len(parts) >= 3 {
								baseURL = parts[2]
							}
							if !seen[name] {
								seen[name] = true
								dep := ExternalDependency{
									Name:    name,
									BaseURL: baseURL,
								}
								ctx.ExternalDeps = append(ctx.ExternalDeps, dep)
							}
						}
					}
				}
			}

			if fn.Body == nil {
				continue
			}

			// Detect HTTP client calls
			if hasHTTP {
				extractHTTPClientCalls(fn.Body, owner, ctx, seen)
			}

			// Detect resty client usage
			if hasResty {
				extractRestyClientCalls(fn.Body, owner, ctx, seen)
			}

			// Detect gRPC connections
			if hasGRPC {
				extractGRPCCalls(fn.Body, owner, ctx, seen)
			}
		}

		// Also check struct fields for client types
		for _, decl := range file.Decls {
			gd, ok := decl.(*ast.GenDecl)
			if !ok {
				continue
			}
			for _, spec := range gd.Specs {
				ts, ok := spec.(*ast.TypeSpec)
				if !ok {
					continue
				}
				st, ok := ts.Type.(*ast.StructType)
				if !ok || st.Fields == nil {
					continue
				}
				structName := pkg + "." + ts.Name.Name

				for _, field := range st.Fields.List {
					if len(field.Names) == 0 {
						continue
					}
					fieldType := fieldTypeString(field.Type)
					isClient := false
					if strings.Contains(fieldType, "http.Client") {
						isClient = true
					}
					if strings.Contains(fieldType, "resty.Client") {
						isClient = true
					}
					if strings.Contains(fieldType, "grpc.ClientConn") {
						isClient = true
					}

					if isClient {
						depName := "external-via-" + ts.Name.Name
						if !seen[depName] {
							seen[depName] = true
							ctx.ExternalDeps = append(ctx.ExternalDeps, ExternalDependency{
								Name:    depName,
								BaseURL: "(detected from HTTP client field in " + structName + ")",
							})
						}
					}
				}
			}
		}
	}
}

// extractHTTPClientCalls looks for http.Get, http.Post, http.NewRequest calls.
//
// @docspec:intentional "Detects standard net/http client calls and extracts HTTP method and URL"
func extractHTTPClientCalls(body *ast.BlockStmt, owner string, ctx *ProcessorContext, seen map[string]bool) {
	ast.Inspect(body, func(n ast.Node) bool {
		call, ok := n.(*ast.CallExpr)
		if !ok {
			return true
		}

		sel, ok := call.Fun.(*ast.SelectorExpr)
		if !ok {
			return true
		}

		ident, ok := sel.X.(*ast.Ident)
		if !ok {
			return true
		}

		if ident.Name != "http" {
			return true
		}

		method := ""
		switch sel.Sel.Name {
		case "Get":
			method = "GET"
		case "Post":
			method = "POST"
		case "Head":
			method = "HEAD"
		case "NewRequest":
			// http.NewRequest(method, url, body)
			if len(call.Args) >= 2 {
				if lit, ok := call.Args[0].(*ast.BasicLit); ok {
					method = strings.Trim(lit.Value, `"`)
				}
			}
		default:
			return true
		}

		if method == "" {
			return true
		}

		// Try to extract URL from arguments
		urlArg := 0
		if sel.Sel.Name == "NewRequest" {
			urlArg = 1
		}

		url := "(dynamic URL)"
		if urlArg < len(call.Args) {
			if lit, ok := call.Args[urlArg].(*ast.BasicLit); ok {
				url = strings.Trim(lit.Value, `"`)
			}
		}

		depName := "http-call-from-" + owner
		if !seen[depName] {
			seen[depName] = true
			dep := ExternalDependency{
				Name: depName,
				Endpoints: []ExternalDependencyEndpoint{
					{
						Method: method,
						Path:   url,
						UsedBy: []string{owner},
					},
				},
			}
			ctx.ExternalDeps = append(ctx.ExternalDeps, dep)
		}

		return true
	})
}

// extractRestyClientCalls looks for resty R().Get/Post/Put/Delete calls.
//
// @docspec:intentional "Detects go-resty fluent chain calls and extracts HTTP method and URL"
func extractRestyClientCalls(body *ast.BlockStmt, owner string, ctx *ProcessorContext, seen map[string]bool) {
	ast.Inspect(body, func(n ast.Node) bool {
		call, ok := n.(*ast.CallExpr)
		if !ok {
			return true
		}

		sel, ok := call.Fun.(*ast.SelectorExpr)
		if !ok {
			return true
		}

		method := ""
		switch sel.Sel.Name {
		case "Get", "Post", "Put", "Delete", "Patch":
			method = strings.ToUpper(sel.Sel.Name)
		default:
			return true
		}

		// Check if this is a resty chain call by looking at the receiver chain
		if isRestyChainCall(sel.X) {
			url := "(dynamic URL)"
			if len(call.Args) > 0 {
				if lit, ok := call.Args[0].(*ast.BasicLit); ok {
					url = strings.Trim(lit.Value, `"`)
				}
			}

			depName := "resty-call-from-" + owner
			if !seen[depName] {
				seen[depName] = true
				dep := ExternalDependency{
					Name: depName,
					Endpoints: []ExternalDependencyEndpoint{
						{
							Method: method,
							Path:   url,
							UsedBy: []string{owner},
						},
					},
				}
				ctx.ExternalDeps = append(ctx.ExternalDeps, dep)
			}
		}

		return true
	})
}

// isRestyChainCall checks if an expression is part of a resty method chain.
//
// @docspec:deterministic
func isRestyChainCall(expr ast.Expr) bool {
	// Look for pattern like client.R().SetXxx()... or resty.New().R()...
	switch e := expr.(type) {
	case *ast.CallExpr:
		if sel, ok := e.Fun.(*ast.SelectorExpr); ok {
			if sel.Sel.Name == "R" || sel.Sel.Name == "SetBody" ||
				sel.Sel.Name == "SetHeader" || sel.Sel.Name == "SetResult" {
				return true
			}
		}
		return isRestyChainCall(e.Fun)
	case *ast.SelectorExpr:
		return isRestyChainCall(e.X)
	}
	return false
}

// extractGRPCCalls looks for grpc.Dial or grpc.NewClient calls.
//
// @docspec:intentional "Detects grpc.Dial/DialContext/NewClient calls and extracts target address"
func extractGRPCCalls(body *ast.BlockStmt, owner string, ctx *ProcessorContext, seen map[string]bool) {
	ast.Inspect(body, func(n ast.Node) bool {
		call, ok := n.(*ast.CallExpr)
		if !ok {
			return true
		}

		sel, ok := call.Fun.(*ast.SelectorExpr)
		if !ok {
			return true
		}

		ident, ok := sel.X.(*ast.Ident)
		if !ok {
			return true
		}

		if ident.Name != "grpc" {
			return true
		}

		if sel.Sel.Name != "Dial" && sel.Sel.Name != "DialContext" && sel.Sel.Name != "NewClient" {
			return true
		}

		target := "(dynamic target)"
		if len(call.Args) > 0 {
			if lit, ok := call.Args[0].(*ast.BasicLit); ok {
				target = strings.Trim(lit.Value, `"`)
			}
		}

		depName := "grpc-" + target
		if !seen[depName] {
			seen[depName] = true
			ctx.ExternalDeps = append(ctx.ExternalDeps, ExternalDependency{
				Name:    depName,
				BaseURL: target,
			})
		}

		return true
	})
}

File: rust/processor/src/extractor/external_dep.rs
//! Detect external HTTP/gRPC dependency patterns in Rust projects.
//!
//! @docspec:module {
//!   id: "docspec-rust-extractor-external-dep",
//!   name: "External Dependency Extractor",
//!   description: "Detects HTTP clients (reqwest, hyper, surf) and gRPC clients (tonic) to extract external service dependencies with URL, method, and endpoint metadata.",
//!   since: "3.0.0"
//! }
//!
//! Looks for:
//! - `reqwest` crate: `Client::new()`, `reqwest::get()`, `Client::get/post/put/delete`
//! - `hyper` client usage
//! - `tonic` gRPC client stubs
//! - `surf` HTTP client
//! - URL string literals that look like external service calls

use serde_json::Value;
use crate::scanner::FileInfo;
use super::{DocSpecExtractor, ProcessorContext};

/// Patterns indicating HTTP client usage.
const HTTP_CLIENT_PATTERNS: &[&str] = &[
    "reqwest::Client",
    "reqwest::get",
    "Client::new()",
    "Client::builder()",
    "hyper::Client",
    "surf::get",
    "surf::post",
    "surf::Client",
];

/// Patterns indicating gRPC client usage (tonic).
const GRPC_CLIENT_PATTERNS: &[&str] = &[
    "tonic::transport::Channel",
    "ServiceClient",
    "connect(",
];

/// @docspec:boundary "External HTTP and gRPC dependency detection from client library usage patterns"
pub struct ExternalDependencyExtractor;

impl DocSpecExtractor for ExternalDependencyExtractor {
    /// @docspec:intentional "Check if any HTTP/gRPC client crate (reqwest, hyper, tonic, surf) is in Cargo.toml"
    fn is_available(&self, ctx: &ProcessorContext) -> bool {
        ctx.has_dependency("reqwest")
            || ctx.has_dependency("hyper")
            || ctx.has_dependency("tonic")
            || ctx.has_dependency("surf")
    }

    /// @docspec:deterministic
    fn extractor_name(&self) -> &'static str {
        "external-dependency"
    }

    /// @docspec:intentional "Detect HTTP and gRPC client usage patterns and extract URL, method, and service metadata"
    fn extract(&self, files: &[FileInfo], ctx: &mut ProcessorContext) {
        let mut dependencies: Vec<Value> = Vec::new();

        for file_info in files {
            let syntax = match syn::parse_file(&file_info.source) {
                Ok(s) => s,
                Err(_) => continue,
            };

            for item in &syntax.items {
                let source = quote::quote!(#item).to_string();

                // Detect HTTP clients
                let has_http_client = HTTP_CLIENT_PATTERNS.iter().any(|p| source.contains(p));
                if has_http_client {
                    let owner = item_name(item, &file_info.module);
                    let urls = extract_url_literals(&source);
                    let methods = extract_http_methods(&source);

                    if urls.is_empty() {
                        // Generic HTTP client detected but no specific URL
                        let dep_name = format!("http-via-{}", owner.split("::").last().unwrap_or(&owner));
                        if !dep_exists(&dependencies, &dep_name) {
                            dependencies.push(serde_json::json!({
                                "name": dep_name,
                                "baseUrl": "(detected from HTTP client field)",
                            }));
                        }
                    } else {
                        for url in &urls {
                            let base_url = extract_base_url(url);
                            let dep_name = url_to_service_name(&base_url);
                            if !dep_exists(&dependencies, &dep_name) {
                                let mut endpoints: Vec<Value> = Vec::new();
                                for method in &methods {
                                    endpoints.push(serde_json::json!({
                                        "method": method,
                                        "path": extract_path_from_url(url),
                                        "usedBy": [owner.clone()],
                                    }));
                                }
                                let mut dep = serde_json::json!({
                                    "name": dep_name,
                                    "baseUrl": base_url,
                                });
                                if !endpoints.is_empty() {
                                    dep.as_object_mut().unwrap()
                                        .insert("endpoints".to_string(), Value::Array(endpoints));
                                }
                                dependencies.push(dep);
                            }
                        }
                    }
                }

                // Detect gRPC clients
                let has_grpc_client = GRPC_CLIENT_PATTERNS.iter().any(|p| source.contains(p));
                if has_grpc_client {
                    let owner = item_name(item, &file_info.module);
                    let urls = extract_url_literals(&source);
                    let base_url = urls.first().cloned()
                        .unwrap_or_else(|| "(detected from gRPC channel)".to_string());

                    // Try to extract service name from tonic-generated client types
                    let service_name = extract_grpc_service_name(&source)
                        .unwrap_or_else(|| format!("grpc-via-{}", owner.split("::").last().unwrap_or(&owner)));

                    if !dep_exists(&dependencies, &service_name) {
                        dependencies.push(serde_json::json!({
                            "name": service_name,
                            "baseUrl": base_url,
                            "protocol": "gRPC",
                            "usedBy": [owner],
                        }));
                    }
                }
            }
        }

        if !dependencies.is_empty() {
            ctx.external_dependencies = dependencies;
        }
    }
}

/// Get a qualified name for an item.
///
/// @docspec:deterministic
fn item_name(item: &syn::Item, module: &str) -> String {
    match item {
        syn::Item::Fn(f) => format!("{}::{}", module, f.sig.ident),
        syn::Item::Struct(s) => format!("{}::{}", module, s.ident),
        syn::Item::Impl(imp) => {
            if let syn::Type::Path(tp) = imp.self_ty.as_ref() {
                let name = tp.path.segments.last()
                    .map(|s| s.ident.to_string())
                    .unwrap_or_default();
                format!("{}::{}", module, name)
            } else {
                module.to_string()
            }
        }
        _ => module.to_string(),
    }
}

/// Extract URL-like string literals from source text.
///
/// @docspec:deterministic
fn extract_url_literals(source: &str) -> Vec<String> {
    let mut urls = Vec::new();
    let prefixes = ["http://", "https://", "grpc://"];

    for prefix in &prefixes {
        let mut search_from = 0;
        while let Some(pos) = source[search_from..].find(prefix) {
            let abs_pos = search_from + pos;
            // Find the end of the URL (next quote, whitespace, or closing paren)
            let url_start = abs_pos;
            let rest = &source[url_start..];
            let end = rest.find(|c: char| c == '"' || c == '\'' || c == ')' || c == ' ' || c == ',')
                .unwrap_or(rest.len());
            let url = rest[..end].to_string();
            if url.len() > prefix.len() {
                urls.push(url);
            }
            search_from = abs_pos + end;
        }
    }

    urls
}

/// Extract HTTP method names from source text.
///
/// @docspec:deterministic
fn extract_http_methods(source: &str) -> Vec<String> {
    let methods_map = [
        (".get(", "GET"),
        (".post(", "POST"),
        (".put(", "PUT"),
        (".delete(", "DELETE"),
        (".patch(", "PATCH"),
        (".head(", "HEAD"),
    ];

    let mut methods = Vec::new();
    for (pattern, method) in &methods_map {
        if source.contains(pattern) && !methods.contains(&method.to_string()) {
            methods.push(method.to_string());
        }
    }
    methods
}

/// Extract the base URL (scheme + host) from a full URL.
///
/// @docspec:deterministic
fn extract_base_url(url: &str) -> String {
    // Find the third slash (after scheme://)
    if let Some(scheme_end) = url.find("://") {
        let rest = &url[scheme_end + 3..];
        if let Some(path_start) = rest.find('/') {
            return url[..scheme_end + 3 + path_start].to_string();
        }
    }
    url.to_string()
}

/// Extract the path portion from a URL.
///
/// @docspec:deterministic
fn extract_path_from_url(url: &str) -> String {
    if let Some(scheme_end) = url.find("://") {
        let rest = &url[scheme_end + 3..];
        if let Some(path_start) = rest.find('/') {
            return rest[path_start..].to_string();
        }
    }
    "/".to_string()
}

/// Convert a base URL to a service name.
///
/// @docspec:deterministic
fn url_to_service_name(base_url: &str) -> String {
    if let Some(scheme_end) = base_url.find("://") {
        let host = &base_url[scheme_end + 3..];
        let host = host.split(':').next().unwrap_or(host);
        let host = host.split('.').next().unwrap_or(host);
        return host.to_string();
    }
    "unknown-service".to_string()
}

/// Try to extract a gRPC service name from tonic client type patterns.
///
/// @docspec:deterministic
fn extract_grpc_service_name(source: &str) -> Option<String> {
    // Look for patterns like `FooServiceClient::` or `FooClient::`
    let client_suffix = "Client::";
    if let Some(pos) = source.find(client_suffix) {
        // Walk backward to find the start of the identifier
        let before = &source[..pos];
        let start = before.rfind(|c: char| !c.is_alphanumeric() && c != '_')
            .map(|p| p + 1)
            .unwrap_or(0);
        let name = &source[start..pos];
        if !name.is_empty() && name != "reqwest" && name != "hyper" && name != "surf" {
            return Some(name.to_string());
        }
    }
    None
}

/// Check if a dependency with the given name already exists.
///
/// @docspec:deterministic
fn dep_exists(deps: &[Value], name: &str) -> bool {
    deps.iter().any(|d| d.get("name").and_then(|n| n.as_str()) == Some(name))
}
